src/components/shared/thread/ActivityGraphFills2.js

```diff
@@ -0,0 +1,846 @@
+      samples.length - 1,
+      prevSampleTime,
+      sampleTime,
+      sampleTime + interval,
+      sampleCPU
+    );
+  }
+
+  /**
+   * Mutate the percentage buffers, by taking this category, and accumulating its
+   * percentage into the buffer.
+   */
+  _accumulateInCategory(
+    category: IndexIntoCategoryList,
+    sampleIndex: IndexIntoSamplesTable,
+    prevSampleTime: Milliseconds,
+    sampleTime: Milliseconds,
+    nextSampleTime: Milliseconds,
+    sampleCPU: number
+  ) {
+    const {
+      rangeEnd,
+      rangeStart,
+      categoryDrawStyles,
+      xPixelsPerMs,
+      canvasPixelWidth,
+    } = this.renderedComponentSettings;
+    if (sampleTime < rangeStart || sampleTime >= rangeEnd) {
+      return;
+    }
+
+    const categoryDrawStyle = categoryDrawStyles[category];
+    const percentageBuffers = this.mutablePercentageBuffers[category];
+
+    if (categoryDrawStyle.selectedFillStyle === 'transparent') {
+      return;
+    }
+
+    const sampleStart = (prevSampleTime + sampleTime) / 2;
+    const sampleEnd = (sampleTime + nextSampleTime) / 2;
+    let pixelStart = (sampleStart - rangeStart) * xPixelsPerMs;
+    let pixelEnd = (sampleEnd - rangeStart) * xPixelsPerMs;
+    pixelStart = Math.max(0, pixelStart);
+    pixelEnd = Math.min(canvasPixelWidth - 1, pixelEnd);
+    const intPixelStart = Math.floor(pixelStart);
+    const intPixelEnd = Math.floor(pixelEnd);
+
+    // For every sample, we have a fractional interval of this sample's
+    // contribution to the graph's pixels.
+    //
+    // v       v       v       v       v       v       v       v       v
+    // +-------+-------+-----+-+-------+-------+-----+-+-------+-------+
+    // |       |       |     |///////////////////////| |       |       |
+    // |       |       |     |///////////////////////| |       |       |
+    // |       |       |     |///////////////////////| |       |       |
+    // +-------+-------+-----+///////////////////////+-+-------+-------+
+    //
+    // We have a device-pixel array of contributions. We map the fractional
+    // interval to this array of device pixels: Fully overlapping pixels are
+    // 1, and the partial overlapping pixels are the degree of overlap.
+
+    //                                 |
+    //                                 v
+    //
+    // +-------+-------+-------+-------+-------+-------+-------+-------+
+    // |       |       |       |///////////////+-------+       |       |
+    // |       |       |       |///////////////////////|       |       |
+    // |       |       +-------+///////////////////////|       |       |
+    // +-------+-------+///////////////////////////////+-------+-------+
+    const percentageBuffer = this._pickPercentageBuffer(
+      percentageBuffers,
+      sampleIndex
+    );
+    for (let i = intPixelStart; i <= intPixelEnd; i++) {
+      percentageBuffer[i] += 1;
+      percentageBuffers.perPixelWeight[i] += sampleCPU;
+      this.mutableTotalWeightgBuffer[i] += sampleCPU;
```

Comment:
What are these two buffers for? I think you don't need them. I think you can use percentageBuffer, like I recommended in the document.